functorch/_src/python_key.py

```diff
@@ -39,51 +48,33 @@ def pythonkey_meta():
         USE_META = False
 
 
-def get_output_device(devices, op):
-    # The device propagation is a bit sketchy.
-    # aten::index(CPU, CUDA) => CPU tensor
-    # aten::index(CUDA, CPU) => CUDA tensor
-    if op == aten.index:
-        return devices[0]
-    devices = list(set(devices))
-    if len(devices) == 1:
-        return devices[0]
-    else:
-        for device in devices:
-            if device.type == 'cuda':
-                return device
-        raise RuntimeError("Couldn't infer output device from input device")
-
-
 class PythonTensor(torch.Tensor):
     elem: torch.Tensor
 
     __slots__ = ['elem', 'proxy']
 
     @staticmethod
-    def __new__(cls, elem, proxy, device=None):
-        # The wrapping tensor (PythonTensor) is just a meta tensor, so it
-        # doesn't hold any memory (meta tensor is generally the preferred type
-        # of tensor you want to make a subclass from)...
-
-        r = torch.Tensor._make_wrapper_subclass(
-            cls, elem.size(),
-            strides=elem.stride(), storage_offset=elem.storage_offset(),
-            dtype=elem.dtype, layout=elem.layout, requires_grad=elem.requires_grad,
-            device=(elem.device if device is None else device),
-        )
-
-        # ...the real tensor is held as an element on the tensor.
-        if USE_META:
-            r.elem = elem.to('meta')
-        else:
-            r.elem = elem
+    def __new__(cls, elem, proxy):
+        # Wrapping something in PythonTensor implicitly detaches
+        # gradients.  If something required grad, we will collect it as if it
+        # were a leaf.  A consequence of detaching in this way is you
+        # need to maintain a parameter cache when translating tensors
+        # into PythonTensor, so you don't create multiple copies of
+        # a gradient (they are aliased, but they would count as independent
+        # leaves).  An alternate strategy would be to avoid implicitly
+        # detaching and instead "catch" gradients as they exit the
+        # PythonTensor boundary.
+        # assert not elem.requires_grad or not torch.is_grad_enabled()
+
+        r = torch.Tensor._make_subclass(cls, elem, elem.requires_grad)
```

Comment:
This doesn't work for meta tensors since at this point, `elem` will be a meta tensor. So we're just gonna make a PythonTensor with the meta device anyways.

That's why I went through all of the shenanigans of inferring the output device - if we run with meta tensors, then at no point do we have the actual output device of the operator. All you have is the device of the input tensors.

So... ripping out the device inference logic will make the meta-tracing stuff not work at all, in which case we should just remove all of it :P

Reply:
This seems like a bigger structural problem for meta tensors. Will need to think about this...